src/Mutant/Generator/MutationsGenerator.php

```diff
@@ -128,8 +130,12 @@ public function generate(bool $onlyCovered, string $filter = ''): array
      */
     private function getMutationsFromFile(SplFileInfo $file, bool $onlyCovered, array $mutators): array
     {
-        /** @var Node[] $initialStatements */
-        $initialStatements = $this->parser->parse($file->getContents());
+        try {
+            /** @var Node[] $initialStatements */
+            $initialStatements = $this->parser->parse($file->getContents());
+        } catch (\Throwable $t) {
+            throw ParserException::fromInvalidFile($file, $t);
```

Comment:
I'm not sure of the flow but if the file here is "just" a PHP file, it's _valid_ for the file to not be a valid PHP file. There can be various reasons for this so instead of bailing out hard like that, I would just ignore the file, eventually log it somewhere

Reply:
In my opinion, if you have an invalid php file, you don't intend infection to make use of that file, so you should add it to an ignore list.

Before, we would error on it as well, the message was just really unclear.

Reply:
Well it's a choice to be made but personally I would much prefer to just have to include `src` and not having to whitelist the couple of invalid files in it than having to whitelist all of them which can be really annoying

Reply:
@infection/core what's your opinion on this? We have the choice of:

- We keep it as it is: if the file cannot be parsed (e.g. it is an invalid PHP file), we bail out (i.e. interrupt the whole process) on this error: no PHP file cannot be invalid
- We accept that a PHP file cannot be valid so in case of failure, we just skip this file

In my experience, there is several cases where I had to deal with invalid PHP files precisely to test something, so the first option is more annoying to use IMO as it requires the user the whitelist all the invalid files. The second solution however wouldn't require the user any extra action.

Second element which is a bit lost in the collapsed discussion: if we go with the second solution, `ParserException` can be completely removed. If we go with the first solution, I do not thing we should have `ParserException` neither:

- It does not bring anything itself: the user should never catch it, so the only thing needed is a self-explanatory message, not a new exception
- Would having an exception preferred, this exception should be an `\InvalidArgumentException` not an `\Exception`

Reply:
I like zero-config style of CLI tools, where you don't need to configure anything to get started.

However, are there any *real* examples of projects where they need to have invalid files?

> I had to deal with invalid PHP files precisely to *test* something

If you had to *test* something, probably it had to be in `./tests` dir, not `./src`? As I understand production-ready project should have only valid files in `src`. Otherwise, how would it work?

----

My 2 cents: 

* IMO, it would be a new enhancement and out of this PR since this one just improves the error message
* if we really need to handle cases for projects with invalid PHP files (but first we should understand how many of them, if any), I'm for ignoring exception and logging/displaying it to the console rather than re-throwing

Also, the original issue (#382) [does not have a clear example](https://github.com/infection/infection/issues/382#issuecomment-396090725) which makes me think it's probably not a real case. Did I miss something?

Reply:
How are that tests passing for such projects while using the very same invalid files?

Reply:
>How are that tests passing for such projects while using the very same invalid files?

It depends a lot on how those invalid files are used. One example could be you expect a certain failure of your application if you try to load it.

https://travis-ci.org/humbug/box/jobs/391429814

>Also, the original issue (#382) does not have a clear example which makes me think it's probably not a real case

It is, I indeed forgot to link the original project which is [Box](https://travis-ci.org/humbug/box/jobs/391429814). Clone and `make tm` will result in this error.

>However, are there any real examples of projects where they need to have invalid files?

I don't have one on top of my head, but I remember facing this issue with PHP-Scoper. That said maybe it was a slightly different case which could be irrelevant here

Reply:
> One example could be you expect a certain failure of your application if you try to load it.

I wonder why would we mutate such files in the first place...

Reply:
>I wonder why would we need to mutate such files in the first place...

We shouldn't, but the question is in terms of usability: do we simply ignore it (not extra action required for the user) or we require the user to whitelist them.

However I think @borNfreee has a point:

>If you had to test something, probably it had to be in ./tests dir, not ./src?

So I guess whitelisting it is fine since it is likely to not affect the users ever.

So that being said, I would come back to the original proposal:

> if we go with the second solution, `ParserException` can be completely removed. If we go with the first solution, I do not thing we should have `ParserException` neither:
>
>- It does not bring anything itself: the user should never catch it, so the only thing needed is a self-explanatory message, not a new exception
>- Would having an exception preferred, this exception should be an `\InvalidArgumentException` not an `\Exception`

Reply:
> It does not bring anything itself: the user should never catch it, so the only thing needed is a self-explanatory message, not a new exception

My reasoning for the `ParserException` is as follows:
* Its 'easier' to place the logic for the creation of the exception in a seperate exception, and
not make it the responsibility of the Generator.
* Being presented with a `InvalidArgumentException` may give the user the idea that they gave some wrong input somewehre.

Reply:
Guys, let's move forward with this simple PR. 

As I understand, we agreed that invalid code should not be in `./src` in most of the cases, in other cases (I guess it is < 1%) users will add an exclusion).

So the open question is just with `ParserException` vs `InvalidArgumentException`, right? 

> Being presented with a `InvalidArgumentException` may give the user the idea that they gave some wrong input somewehre.

I find this reasonable. If anyone still has objections regarding current state of the PR - please comment. If not - let's merge it.

Reply:
Being presented with any kind of exception is bad, no matter how aptly named it is. 

ParserException is certainly better named, and best the enemy of the good, so I'm all in for the new ParserException if it can include the original exception it was caused by. It'll help debugging problems.